### DISCUSSION TITLE
Stable release

### DISCUSSION THREAD

**Discussion by sampsapursiainen** [sampsapursiainen/zeffiro_interface#168] [General]
(Check the Milestones for the DLs)!

We would need to release a stable version where the recent contributions have been debugged and some example scripts for understanding the usage would be provided.

The parts to be checked include: 
- Finite element mesh creator
- Stimulation tool
- Inverse source localization scripts (including Kalman filter) 
- Deployment tests and usage examples would be provided in tests and scripts folder, respectively.

I would request that each contributor would (1) debug tools they have developed, (2) that the tools can be opened using zef_tool_start.m function, (3) check that those can be run both in graphical user interface mode and on command line in terminal (without the java interface or any graphics available), and (4) The scripts folder includes a self-supported demonstrator that (a) starts the interface, (b) imports or opens a project, (c) runs the tool with test data (numerical simulations if available, a small set can be uploaded in the repository, if needed and if it is ok to do so regaring any IPRs, alternatively an online dataset can be downloaded for the deployment run, if it is not applicable otherwise).

**Comment by SeSodesa** [comment]
Once the stable release is made, I would suggest bumping the version to `6.0.0`, just because semantic versioning has not been properly followed thus far. From then on, the guidelines listed on <https://semver.org/> should be followed:

> Given a version number MAJOR.MINOR.PATCH, increment the:
>
> 1. MAJOR version when you make [backwards] incompatible API changes,
> 2. MINOR version when you add functionality in a backwards compatible manner,
> 3. PATCH version when you make backwards compatible bug fixes
>
> Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.

Possible followers of the project on MathWorks File Exchange probably expect proper semantic versioning to be followed, as MathWorks seems to require its use when projects are published and versions released.

**Comment by SeSodesa** [comment]
@sampsapursiainen Could you by any means list the names, types and possible values of the fields of `zef` needed in mesh generation and lead field construction here (since you are the only one who *really* knows these)? I could then document these in the relevant functions in the code.

**Comment by SeSodesa** [comment]
@JoonasJL @GitFernandoGalaz @Yusuf-Oluwatoki If you guys know the fields used by the "published" inverse tools, it would also help if these were listed in an open manner (see the previous message in this discussion).

**Reply by JoonasJL** [reply]
I am about to reformalize all tools (ClassicalSparseMethods, EXP, and SESAME) completely owned by me to have their own field named "parameters", where all the tool specific parameters are. This also helps me to use more descriptive variable names. Top of that, I would like to edit Dipole scan so that it would be command line friendly. Currently it is not.

Because I am about to do a reformation this big, this would be extremely good opportunity to suggest new practices.

I will give you beamformer, IAS, RAMUS fields and GMM v2 (even if it is not an inverse method) in the end of the week if I end up being quick with an exercise checking duty.

**Reply by SeSodesa** [reply]
You could enforce the presence of the parameters if you wrote a (handle) class, let's call it `InverseParameters`, to store the fields, instead of using a plain old `struct`. You could even set sensible default values for the parameters with class property validation.

**Reply by SeSodesa** [reply]
@JoonasJL Writing the property structure as a class also allows you to document the code in an accessible way. The class definition

	classdef MyDocumentedClass

		%
		% MyDocumentedClass
		%
		% The description of this very well documented class. Could also
		% describe the public interface here.
		%

		properties

			%
			% some_index_thing
			%
			% This prop does something related to indexing, and hence it must
			% be a positive integer.
			%
			some_index_thing (1,1) double { mustBeInteger, mustBePositive } = 1

		end % properties

	end % classdef

makes it possible to call

	>> help MyDocumentedClass.some_index_thing

on the Matlab command line to get the help text written above the property in the `properties` block. **Note** that the last `%` sign of the documentation comment needs to touch the property name below, as in be located just above the property for the comment to actually register as a documentation comment. With the top level class documentation comment, this restriction does not seem to be in place.

**Reply by JoonasJL** [reply]
Tell me if I am wrong, but in the sake of consistency, I need the thing called "parameters" for multiple methods. So, what I need to do is write a bunch of classes called EXP_InverseParameters for example and then I make a class object as:
zef.EXP.parameters = EXP_InverseParameters;

Note that I will have the app itself as one field of EXP that is class (or superclass even?) of its own.

**Reply by SeSodesa** [reply]
> Tell me if I am wrong, but in the sake of consistency, I need the thing called "parameters" for multiple methods. So, what I need to do is write a bunch of classes called EXP_InverseParameters for example and then I make a class object as:
zef.EXP.parameters = EXP_InverseParameters;

If the different methods use the same parameters, then you need just one class definition, and each app / method can use that set of parameters. If all of the methods share some parameters, but have some special ones not used by others, you can inherit from the common parameters to avoid code repetition:

	classdef CommonParameters

		%
		% CommonParameters
		%
		% Common parameters used by inverse methods.
		%

		properties

			%
			% common_parameter
			%
			% This is a parameter that all inverse methods use.
			%
			common_parameter

		end % properties

		methods

			function self = CommonParameters(common_parameter)

				%
				% CommonParameters.CommonParameters
				%
				% The constructor of this superclass.
				%

				self.common_parameter = common_parameter;

			end % function

		end % methods

	end % classdef

	classdef SpecificMethodParameters < CommonParameters

		%
		% SpecificMethodParameters < CommonParameters
		%
		% Extends the common set of parameters by parameters needed by a
		% specific inverse method.
		%

		properties

			%
			% specific_parameter
			%
			% This is an additional parameter needed by a specific inverse
			% method.
			%
			specific_parameter

		end % properties

		methods

			function self = SpecificMethodParameters(common_parameter, specific_parameter)

				%
				% SpecificMethodParameters.SpecificMethodParameters
				%
				% The constructor of this subclass.
				%

				% First call the superclass constructor to set its properties.

				self = self@CommonParameters(common_parameter);

				% Then initialize own fields.

				self.specific_parameter = specific_parameter;

			end % function

		end % methods

	end % classdef

Matlab compiles the above into a single type definition for you, something like

	classdef SpecificMethodParameters

		properties

			common_parameter

			specific_parameter

		end

		methods

			function self = SpecificMethodParameters(common_parameter, specific_parameter)

				%
				% SpecificMethodParameters.SpecificMethodParameters
				%
				% The constructor of this subclass.
				%

				% The superclass property is inherited, so simply set it here as if it was my own property.

				self.common_parameter = common_parameter;

				% Then initialize own fields.

				self.specific_parameter = specific_parameter;

			end % function

		end % methods

	end

If you had defined more methods in the base- or superclass `CommonParameters`, those methods would also be available in the inheriting subclass `SpecificMethodParameters`. Although it has to be said that if the base class was abstract, this becomes more complicated.

> Note that I will have the app itself as one field of EXP that is class (or superclass even?) of its own.

If I read this correctly, you are slightly mixing the idea of inheritance vs. composition. You could certainly use composition instead of inheritance (automatic property and methods addition to subclasses), by defining a `CommonProperties` class and then inserting that as one property of `SpecialMethodProperties`. But this does not allow you access the fields as easily without defining additional access methods. For example, you would have to do

	%% Composition use example.

	% Initialize a common parameter object.

	common = CommonParameters(common_parameter);

	% Initialize specific parameter object by inserting the entire common object into it.

	specific = SpecificMethodParameters(common, specific_parameter);

	% Access the properties.

	specific.common.common_parameter

	specific.specific_parameter

instead of

	%% Inheritance use example.

	% Just initialize a specific parameter object with the necessary properties.

	specific = SpecificMethodParameters(common_parameter, specific_parameter);

	% Access the properties.

	specific.common_parameter

	specific.specific_parameter

So inheritance might simplify you code a bit when you actually use the classes you have defined, although not necessarily. It can be abused in certain ways.

**Reply by SeSodesa** [reply]
As for the app, you could have a GUI class, that holds onto the app class (pseudocode):

```matlab
classdef SpecificInverter

	properties % with validation and documentation, hopefully.

		property_1 = default_value_1

		property_2 = default_value_2

		...

	end

	methods

		function self = SpecificInverter(property_1, property_2)

			%
			% SpecificInverter
			%
			% Constructor of this class.
			%

			self.property_1 = property_1;

			self.property_2 = property_2;

			...

		end

		function reconstruction = invert(self)

			% Function code here.

		end

	end

end

classdef InverterGUI < uifigure

	% A GUI that holds onto an inverter in its properties. When this this is
	% started, it starts the inverter in its constructor with the given
	% parameters.

	properties

		%
		% An inverter back-end
		%
		backend (1,1) SpecificInverter

		%
		% A button in this GUI
		%
		button (1,1) matlab.uicontrol.Button

		% Add other graphical components here.

	end

	methods

		function self = InverterGUI(args)

			%
			% InverterGUI
			%
			% Constructor that allows starting the application via a GUI, with
			% name–value arguments.
			%

			arguments

				args.property_1 = default_value_1

				args.property_2 = default_value_2

			end

			self.backend = SpecificInverter(args.property_1, args.property_2);

			self.button = uibutton();

			self.button.Text = string(self.backend.property_1);

			...

		end

		% TODO: GUI callback functions triggered by buttons etc. here. The
		% Start-button would call the invert-method of the back-end, for
		% example. When changing a text field, a specific property of
		% self.backend could be set.

	end

end
```

**Reply by SeSodesa** [reply]
@JoonasJL So in short, yes, if `EXP` is the main GUI or graphical user interface class of your application, then it should have an instance of the class that contains the needed parameters as one of its properties. I'd say inheritance is not the right approach to take in terms of including the parameters into the GUI class, because the GUI is not supposed to be a subtype or -class of the back-end. It is simply a separate object that simply provides a view into the capabilities of the back-end object, which it can do by simply having a "handle" to the object in its properties.

**Reply by JoonasJL** [reply]
Thanks a lot for extensive examples and explanations! That's helps a lot.
What we definitely need then, is the class for ZI's common inverse parameters for starters. I paste them here also if it helps someone:
_inv_sampling_frequency
inv_low_cut_frequency
inv_high_cut_frequency
inv_snr
inv_prior_over_measurement_db
inv_amplitude_db
inv_time_1
inv_time_2
inv_time_3
number_of_frames
inv_data_segment
normalize_data

I don't know, does any other sector of ZI use these:
source_directions
source_direction_mode
source_interpolation_ind
source_positions_

In the current situation, the "methods" that have more shared parameters than these are usually scripted to the same solver as _flavours_. So all we need is love and those parameters in a CommonParameters class.


About apps, I wouldn't redefine the app functionalities in a class. I use appDesigner to make the apps and it creates a class of some kind on its own. What would be nice is to have the app as a property of a class like:
`classdef EXPmethod
properties
app
common_parameters
specific_parameters
end
end`
(I don't know what this code will look like because I didn't get anysort of view)

**Reply by SeSodesa** [reply]
You could just write the GUI in the App Designer, and other code like the inverse functions that need to be directly accessible from the command line without starting the GUI as regular `.m` files. Those should be accessible from within the App Designer view, just like you can call regular Matlab functions like `sin` inside a callback function of an App Designer application, as long as the `.m` files can be found on the Matlab path.

The back-end of the GUI written in App Designer would then be a class in a `.m` file, that inherits from the `CommonParameters` class, as in

    BackEnd < CommonParameters

so the `properties` of `CommonParameters` get automatically added to the properties of `BackEnd` without you having to rewrite them there. The inverse method of this back-end inverter could then be a function in the `methods` block of `BackEnd`, or if the function needs to be in a separate file, the `BackEnd` method could just call the function with parameters stored in the `properties` of `BackEnd`. The file division would be something like

    CommonParameters.m
    BackEnd.m
    inverse_method.m
    GUI.mlapp

The first 2 files contain `classdef`s, the third (if needed as a standalone function) a `function` definition and the 4th the GUI `classdef` written in the App Designer. In the `GUI`, you would have a property (pseudocode)

    GUI.back_end (1,1) BackEnd

with maybe a different name. `BackEnd` does not need to (and *should not*) know about `GUI` as a property to be able to function. We can discuss later how a back-end might send messages to the GUI without GUI being a property of `BackEnd`, if needed.

**Reply by frank-neugebauer** [reply]
At this point, you should also decouple the data processing from the inverse solvers, there is a lot of redundancy in the code.

Decoupling the GUI from the backend makes sense, model-view-controller is also common in web applications, as I had to learn :)
The dipole scan and the dataBank are already designed to work independently of the GUI, also their use can be a little clunky.
This was important for me, because Matlab does not save the GUI objects.

**Reply by SeSodesa** [reply]
Another aspect of this is that if you implement the inverter back-ends and/or properties as classes, you can customize how they are saved to the disk by Matlab ([link][saveloadobj]).

[saveloadobj]: https://se.mathworks.com/help/matlab/save-and-load.html

**Reply by sampsapursiainen** [reply]
Yes, now the data structures of most "inverters" are a hybrids between the window class and zef, but indeed the window class principally should be enough.

**Comment by sampsapursiainen** [comment]
Hi Frank, 

I agree. You contributed this topic importantly by writing those filtering, lead field processing and postprocessing functions that are used by most (?) of the inverse tools at the moment. Perhaps we could try to extend that approach by writing some block to steer the filtering parameters and options which could be accessed by those functions. That would enable selecting common default filter parameters for all tools. There are some parameters which might be made common, whereas now they are not necessarily. One of those is SNR. 

Indeed, the dataBank has a great expandable design, and it is important, that it can be regarded as a separate field in the zef data structure. After a minor modification, dataBank is now started by a script that allows it to be either visible (with GUI) or invisible (without GUI).
